File: packages/gavio-py/gavio/interceptors/guardrails/validators/license.py
"""License / copyright detection validator (F-QUA-10).

Flags known open-source license text (MIT, Apache-2.0, GPL-2.0/3.0,
BSD-3-Clause, MPL-2.0) in a model response before it lands in user code.
Matches against a shipped corpus of hashed 8-word shingles -- no license text
is ever bundled. Detections surface in the guardrail outcome and, via the
guardrails interceptor, in the audit record.
"""

from __future__ import annotations

import hashlib

from ..validator import OutputValidator, ValidationResult
from ._license_fingerprints import LICENSE_FINGERPRINTS

_SHINGLE_N = 8


def _normalize_tokens(text: str) -> list[str]:
    """ASCII-lower-alnum tokeniser -- must stay byte-identical across all SDKs."""
    out: list[str] = []
    cur: list[str] = []
    for ch in text:
        o = ord(ch)
        if 65 <= o <= 90:
            cur.append(chr(o + 32))
        elif (97 <= o <= 122) or (48 <= o <= 57):
            cur.append(ch)
        elif cur:
            out.append("".join(cur))
            cur = []
    if cur:
        out.append("".join(cur))
    return out


def _shingle_hashes(tokens: list[str]) -> set[str]:
    hashes: set[str] = set()
    for i in range(len(tokens) - _SHINGLE_N + 1):
        gram = " ".join(tokens[i : i + _SHINGLE_N])
        hashes.add(hashlib.sha256(gram.encode("utf-8")).hexdigest()[:16])
    return hashes


def detect_licenses(
    content: str,
    licenses: list[str] | None = None,
    min_matches: int = 1,
) -> list[str]:
    """Return the sorted SPDX ids whose fingerprint appears in ``content``."""
    present = _shingle_hashes(_normalize_tokens(content))
    ids = licenses if licenses is not None else list(LICENSE_FINGERPRINTS)
    found: list[str] = []
    for spdx_id in ids:
        fingerprints = LICENSE_FINGERPRINTS.get(spdx_id)
        if not fingerprints:
            continue
        hits = sum(1 for h in fingerprints if h in present)
        if hits >= min_matches:
            found.append(spdx_id)
    return sorted(found)


class LicenseDetectorValidator(OutputValidator):
    """Fails if the content contains recognisable license text (F-QUA-10)."""

    def __init__(
        self,
        licenses: list[str] | None = None,
        min_matches: int = 1,
    ) -> None:
        self._licenses = licenses
        self._min_matches = min_matches

    @property
    def name(self) -> str:
        return "license_detector"

    def detect(self, content: str) -> list[str]:
        return detect_licenses(content, self._licenses, self._min_matches)

    def validate(self, content: str) -> ValidationResult:
        found = self.detect(content)
        if not found:
            return ValidationResult.passed()
        return ValidationResult.failed(f"license text detected: {', '.join(found)}")

File: packages/gavio-js/src/interceptors/guardrails/validators/license.ts
/**
 * License / copyright detection validator (F-QUA-10).
 *
 * Flags known open-source license text (MIT, Apache-2.0, GPL-2.0/3.0,
 * BSD-3-Clause, MPL-2.0) in a model response before it lands in user code.
 * Matches against a shipped corpus of hashed 8-word shingles — no license text
 * is ever bundled. Detections surface in the guardrail outcome and, via the
 * guardrails interceptor, in the audit record.
 */

import { createHash } from 'node:crypto'
import { failed, passed, type OutputValidator, type ValidationResult } from '../validator.js'
import { LICENSE_FINGERPRINTS } from '../data/license-fingerprints.js'

const SHINGLE_N = 8

/** ASCII-lower-alnum tokeniser — must stay byte-identical across all SDKs. */
function normalizeTokens(text: string): string[] {
  const out: string[] = []
  let cur = ''
  for (let i = 0; i < text.length; i++) {
    const c = text.charCodeAt(i)
    if (c >= 65 && c <= 90) cur += String.fromCharCode(c + 32)
    else if ((c >= 97 && c <= 122) || (c >= 48 && c <= 57)) cur += String.fromCharCode(c)
    else if (cur) {
      out.push(cur)
      cur = ''
    }
  }
  if (cur) out.push(cur)
  return out
}

function shingleHashes(tokens: string[]): Set<string> {
  const hashes = new Set<string>()
  for (let i = 0; i + SHINGLE_N <= tokens.length; i++) {
    const gram = tokens.slice(i, i + SHINGLE_N).join(' ')
    hashes.add(createHash('sha256').update(gram, 'utf8').digest('hex').slice(0, 16))
  }
  return hashes
}

export interface LicenseDetectorOptions {
  /** SPDX ids to check; defaults to the whole shipped corpus. */
  licenses?: string[]
  /** Minimum distinct shingle hits before a license is flagged (default 1). */
  minMatches?: number
}

/** Returns the sorted SPDX ids whose fingerprint appears in `content`. */
export function detectLicenses(content: string, options: LicenseDetectorOptions = {}): string[] {
  const { licenses, minMatches = 1 } = options
  const present = shingleHashes(normalizeTokens(content))
  const ids = licenses ?? Object.keys(LICENSE_FINGERPRINTS)
  const found: string[] = []
  for (const id of ids) {
    const fingerprints = LICENSE_FINGERPRINTS[id]
    if (!fingerprints) continue
    let hits = 0
    for (const h of fingerprints) {
      if (present.has(h)) {
        hits++
        if (hits >= minMatches) break
      }
    }
    if (hits >= minMatches) found.push(id)
  }
  return found.sort()
}

/** Fails if the content contains recognisable license text (F-QUA-10). */
export function licenseDetector(options: LicenseDetectorOptions = {}): OutputValidator {
  return {
    name: 'license_detector',
    validate(content: string): ValidationResult {
      const found = detectLicenses(content, options)
      if (found.length === 0) return passed()
      return failed(`license text detected: ${found.join(', ')}`)
    },
  }
}
